clash-lib/src/Clash/Netlist/Util.hs

```diff
@@ -1063,28 +1105,28 @@ mkInput pM = case pM of
         Product _ _ hwtys -> do
           arguments <- zipWithM appendIdentifier (map (i',) hwtys) [0..]
           (ports,_,exprs,_) <- unzip4 <$> mapM (mkInput Nothing) arguments
-          case exprs of
-            [expr] ->
-              let netdecl  = NetDecl Nothing i' hwty
-                  dcExpr   = expr
-                  netassgn = Assignment i' expr
-              in  return (concat ports,[netdecl,netassgn],dcExpr,i')
-            _ ->
-              let netdecl  = NetDecl Nothing i' hwty
-                  dcExpr   = DataCon hwty (DC (hwty,0)) exprs
-                  netassgn = Assignment i' dcExpr
-              in  if null attrs then
-                    return (concat ports,[netdecl,netassgn],dcExpr,i')
-                  else
-                    throwAnnotatedSplitError $(curLoc) "Product"
+          let netdecl  = NetDecl Nothing i' hwty
+              dcExpr   = DataCon hwty (DC (hwty,0)) exprs
+              netassgn = Assignment i' dcExpr
 
-        _ -> return ([(i',hwty)],[],Identifier i' Nothing,i')
+          if null attrs then
+            return (concat ports,[netdecl,netassgn],dcExpr,i')
+          else
+            throwAnnotatedSplitError $(curLoc) "Product"
 
+        _ -> do
+          (decls, i'', expr, hwty1) <- fromPrimitiveType i' hwty
+          return ([(i', hwty1)], decls, expr, i'')
 
     -- PortName specified by user
-    go' (PortName p) (i,hwty) = do
-      pN <- uniquePortName p i
-      return ([(pN,hwty)],[],Identifier pN Nothing,pN)
+    go'
+      :: PortName
+      -> (Identifier, HWType)
+      -> NetlistMonad ([(Identifier, HWType)], [Declaration], Expr, Identifier)
+    go' (PortName p) (i0, hwty0) = do
+      pN <- uniquePortName p i0 -- << Seems silly, remove?
```

Comment:
No, don't remove. When `p` is not the empty string, it ensures that we don't pick another basic identifier with a differently cased letters in a case-insensitive language. So it doesn't actually change `p` when `p` is not empty, but does ensure it's unique w.r.t. generated identifiers.

Reply:
Got it, forgot to remove that comment after our discussion yesterday.